### PR TITLE
Kh/dc,csv download

### LINKED ISSUE #49: Add CSV download button to get results in CSV format
# User Story

As a user, from any courses search, I can press a "download as CSV" button, and get the results of that search in CSV format.

# Acceptance Criteria

- [x] There should be a "CSV Download" button on the webpage next to the submit button.
- [x] When this button is clicked we should get the results in CSV format.
- [x] Error message if the search fails or does not return anything.

# Implementation Todos

Front end:

- [x] Button displayed on-page which links to an endpoint which returns back a CSV file.

Back end:

- [x] Add a class ObjectToCSVService which takes the data stored in the course repository and returns a CSV file to the controller.
- [x] Add a controller returns the CSV file to the frontend.

Testing:

- [x] Front end tests pass and there is adequate coverage
- [x] Back end tests pass and there is adequate coverage

### PR DESCRIPTION
**Pull request by kunalhandaUCSB**
This pull request closes #49. This branch adds a csv download button which downloads a csv file containing the result of the course search.

Before, no CSV Download button:

image

After: CSV Download Button

course-search-csv-after

### CODE REVIEW

**Comment by kunalhandaUCSB** [comment]
@scottpchow23 @tkomarlu @aezhao065 Fixed this pull request, please re-review

**Review by aezhao065 — DISMISSED** [review]

File: javascript/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
+    // const downloadCSVurl = () => {
+        
+    //     const url = `/api/public/basicsearch/downloadcsv?qtr=${quarter}&dept=${department}&level=${level}`;
+    //     return url;
+    // }
+    // const handleDownload = (event) => {
+    //     event.preventDefault();
+    //     window.location.href=downloadCSVurl();
+    // }
+
```
Comment (andrewberry391) [inline javascript/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:19]: Is it necessary to keep this commented code in the final

File: javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js [hunk 2]
```diff
@@ -82,5 +82,55 @@ describe("BasicCourseSearchForm tests", () => {
...
+        await waitFor(() => expect(setCourseJSONSpy).toHaveBeenCalledTimes(1));
+        await waitFor(() => expect(fetchJSONSpy).toHaveBeenCalledTimes(1));
+
+        // assert that ourSpy was called with the right value
+        expect(setCourseJSONSpy).toHaveBeenCalledWith(sampleReturnValue);
+        expect(fetchJSONSpy).toHaveBeenCalledWith(expect.any(Object), expectedFields);
+
+        // lets try the download csv now
+        // const csvButton = getByText("Download CSV");
+        // userEvent.click(csvButton);
```
Comment (andrewberry391) [inline javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js:133]: Is it necessary to keep this commented code in the final

File: src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java [hunk 3]
```diff
@@ -1,24 +1,27 @@
...
 import org.springframework.web.bind.annotation.GetMapping;
-
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
 
 import edu.ucsb.courses.services.UCSBCurriculumService;
 
+// import edu.ucsb.courses.services.DownloadCsvService;
+
```
Comment (andrewberry391) [inline src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java:24]: Is it necessary to keep this commented code in the final

File: src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java [hunk 4]
```diff
@@ -30,15 +33,29 @@
...
+    //     String filename = String.format("courses-%s-%s-%s.csv", qtr,dept,level);
+    //     response.setContentType("text/csv");
+    //     response.setHeader("Content-Disposition", "attachment; filename=\"" + filename + "\"");
+    //     try {
+    //         downloadCsvService.listToCSV(coursePageJson, response.getWriter());
+    //     } catch (IOException e) {
+    //         logger.error("Error Writing to Response Stream{}", e);
+    //     }
+    // }     
+}
```
Comment (andrewberry391) [inline src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java:61]: Is it necessary to keep this commented code in the final

File: src/main/java/edu/ucsb/courses/services/DownloadCsvService.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+//             try {
+//                 //fileWriter.flush();
+//                 //fileWriter.close();
+//                 csvPrinter.close();
+//             } catch (IOException e) {
+//                 logger.error("Flushing/closing error!{}", e);
+//             }
+//         }
+//     }
+// }
```
Comment (andrewberry391) [inline src/main/java/edu/ucsb/courses/services/DownloadCsvService.java:65]: Is it necessary to keep this commented code in the final

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java [hunk 6]
```diff
@@ -1,16 +1,20 @@
 package edu.ucsb.courses.controllers;
 
 import edu.ucsb.courses.config.SecurityConfig;
+import edu.ucsb.courses.documents.Course;
+import edu.ucsb.courses.documents.CoursePage;
+// import edu.ucsb.courses.services.DownloadCsvService;
```
Comment (andrewberry391) [inline src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:6]: Is it necessary to keep this commented code in the final

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java [hunk 7]
```diff
@@ -28,12 +37,20 @@
...
+    private final Logger logger = LoggerFactory.getLogger(BasicSearchControllerTests.class);
+
     @Autowired
     private MockMvc mockMvc;
 
     @MockBean
     private UCSBCurriculumService ucsbCurriculumService;
 
+    // @MockBean
+    // private DownloadCsvService downloadCsvService;
```
Comment (andrewberry391) [inline src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:49]: Is it necessary to keep this commented code in the final

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java [hunk 8]
```diff
@@ -50,4 +67,27 @@ public void test_basicSearch() throws Exception {
...
+    //     String url = String.format(urlTemplate, "20204", "CMPSC", "L");
+    //     when(ucsbCurriculumService.getJSON(any(String.class), any(String.class), any(String.class)))
+    //             .thenReturn(cpJSON);
+    //     // when(downloadCsvService.listToCSV(any(String.class), any(PrintWriter.class)));
+    //     MvcResult response = mockMvc.perform(get(url).contentType("text/csv")).andExpect(status().isOk())
+    //             .andReturn();
+    //     String responseString = response.getResponse().getContentAsString();
+    //     assertEquals("", responseString);
+    // }
+
```
Comment (andrewberry391) [inline src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:92]: Is it necessary to keep this commented code in the final

**Review by andrewberry391 — CHANGES_REQUESTED** [review]
Looks great other than some commented code that is probably not necessary for the final version.

**Review by andrewberry391 — DISMISSED** [review]

File: javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js (hunk 2 above)
Comment (pconrad) [reply javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js:133]: I agree with Andrew.. this should be deleted

File: src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java (hunk 3 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java:24]: ^^^ What he said

File: src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java (hunk 4 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/courses/controllers/BasicSearchController.java:61]: ^^^ What he said

File: src/main/java/edu/ucsb/courses/services/DownloadCsvService.java (hunk 5 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/courses/services/DownloadCsvService.java:65]: ^^^ What he said...

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java (hunk 6 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:6]: ^^^ What he said

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java (hunk 7 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:49]: ^^^ What he said

File: src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java (hunk 8 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/courses/controllers/BasicSearchControllerTests.java:92]: ^^^ What he said

**Review by pconrad — CHANGES_REQUESTED** [review]
Please remove the commented out code per Andrew's code review :-)

(If he had not pointed it out, I would have)

**Comment by kunalhandaUCSB** [comment]
@pconrad @andrewberry391 @scottpchow23 @tkomarlu deleted unnecessary comments, please re-review

File: javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js (hunk 2 above)
Comment (pconrad) [reply javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js:133]: This code still seems to be here...

Comment (pconrad) [reply javascript/src/test/components/BasicCourseSearch/BasicCourseSearchForm.test.js:133]: But ok...

**Review by pconrad — APPROVED** [review]
Ok, LGTM.   I will wait to test it on qa in the morning before merging to main

**Review by maodougansidui — COMMENTED** [review]
just reviewed.

**Comment by pconrad** [comment]
40 pts

**Pull request merged**